File: CS/Q577904/CF_HTMLHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.Commands;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.Office.Utils;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraRichEdit.Commands.Internal;
using DevExpress.XtraRichEdit.Export.Html;

namespace Q577904
{
    public class CF_HTMLHelper
    {
        //HTML Clipboard Format http://msdn.microsoft.com/en-us/library/aa767917(v=vs.85).aspx
        const string StartFragmentTag = "<!--StartFragment-->";
        const string EndFragmentTag = "<!--EndFragment-->";

        const string bodyTag = "<body>\r\n";
        const string bodyTagClose = "</body>";
        const string EmptyDescription = "Version:0.9\r\nStartHTML:{0:D10}\r\nEndHTML:{1:D10}\r\nStartFragment:{2:D10}\r\nEndFragment:{3:D10}\r\n";

        public static string GetHtmlClipboardFormat(string html)
        {
            int startBodyTagPos = html.IndexOf(bodyTag);
            int bodyEndTagPos = html.LastIndexOf(bodyTagClose);

            int contentBeforeFramentLength = startBodyTagPos + bodyTag.Length;
            string contentBeforeFragment = html.Substring(0, contentBeforeFramentLength);

            string fragment = html.Substring(contentBeforeFramentLength, bodyEndTagPos - contentBeforeFramentLength);

            string contentAfterFragment = html.Substring(bodyEndTagPos, html.Length - bodyEndTagPos);

            string result = Get_CF_HTML(contentBeforeFragment + StartFragmentTag, fragment, EndFragmentTag + contentAfterFragment);

            return result;
        }

        static string Get_CF_HTML(string contentBeforeFragment, string fragment, string contentAfterFragment)
        {
            int contentBeforeFragmentCount = Encoding.UTF8.GetByteCount(contentBeforeFragment);
            int fragmentCount = Encoding.UTF8.GetByteCount(fragment);
            int contentAfterFragmentCount = Encoding.UTF8.GetByteCount(contentAfterFragment);

            int descriptionOffset = Encoding.UTF8.GetByteCount(String.Format(EmptyDescription, 0, 0, 0, 0));
            int endHTMLOffset = descriptionOffset + contentBeforeFragmentCount + fragmentCount + contentAfterFragmentCount;
            int startFragmentOffset = descriptionOffset + contentBeforeFragmentCount;
            int endFragmentOffset = descriptionOffset + contentBeforeFragmentCount + fragmentCount;

            string description = String.Format(EmptyDescription, descriptionOffset, endHTMLOffset, startFragmentOffset, endFragmentOffset);

            StringBuilder content = new StringBuilder();
            content.Append(description);
            content.Append(contentBeforeFragment);
            content.Append(fragment);
            content.Append(contentAfterFragment);
            return content.ToString();
        }
    }
}

File: VB/Q577904/CF_HTMLHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.Commands
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.Office.Utils
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraRichEdit.Commands.Internal
Imports DevExpress.XtraRichEdit.Export.Html

Namespace Q577904
    Public Class CF_HTMLHelper
        'HTML Clipboard Format http://msdn.microsoft.com/en-us/library/aa767917(v=vs.85).aspx
        Private Const StartFragmentTag As String = "<!--StartFragment-->"
        Private Const EndFragmentTag As String = "<!--EndFragment-->"

        Private Const bodyTag As String = "<body>" & ControlChars.CrLf
        Private Const bodyTagClose As String = "</body>"
        Private Const EmptyDescription As String = "Version:0.9" & ControlChars.CrLf & "StartHTML:{0:D10}" & ControlChars.CrLf & "EndHTML:{1:D10}" & ControlChars.CrLf & "StartFragment:{2:D10}" & ControlChars.CrLf & "EndFragment:{3:D10}" & ControlChars.CrLf

        Public Shared Function GetHtmlClipboardFormat(ByVal html As String) As String
            Dim startBodyTagPos As Integer = html.IndexOf(bodyTag)
            Dim bodyEndTagPos As Integer = html.LastIndexOf(bodyTagClose)

            Dim contentBeforeFramentLength As Integer = startBodyTagPos + bodyTag.Length
            Dim contentBeforeFragment As String = html.Substring(0, contentBeforeFramentLength)

            Dim fragment As String = html.Substring(contentBeforeFramentLength, bodyEndTagPos - contentBeforeFramentLength)

            Dim contentAfterFragment As String = html.Substring(bodyEndTagPos, html.Length - bodyEndTagPos)

            Dim result As String = Get_CF_HTML(contentBeforeFragment & StartFragmentTag, fragment, EndFragmentTag & contentAfterFragment)

            Return result
        End Function

        Private Shared Function Get_CF_HTML(ByVal contentBeforeFragment As String, ByVal fragment As String, ByVal contentAfterFragment As String) As String
            Dim contentBeforeFragmentCount As Integer = Encoding.UTF8.GetByteCount(contentBeforeFragment)
            Dim fragmentCount As Integer = Encoding.UTF8.GetByteCount(fragment)
            Dim contentAfterFragmentCount As Integer = Encoding.UTF8.GetByteCount(contentAfterFragment)

            Dim descriptionOffset As Integer = Encoding.UTF8.GetByteCount(String.Format(EmptyDescription, 0, 0, 0, 0))
            Dim endHTMLOffset As Integer = descriptionOffset + contentBeforeFragmentCount + fragmentCount + contentAfterFragmentCount
            Dim startFragmentOffset As Integer = descriptionOffset + contentBeforeFragmentCount
            Dim endFragmentOffset As Integer = descriptionOffset + contentBeforeFragmentCount + fragmentCount

            Dim description As String = String.Format(EmptyDescription, descriptionOffset, endHTMLOffset, startFragmentOffset, endFragmentOffset)

            Dim content As New StringBuilder()
            content.Append(description)
            content.Append(contentBeforeFragment)
            content.Append(fragment)
            content.Append(contentAfterFragment)
            Return content.ToString()
        End Function
    End Class
End Namespace
